trinity/tx_pool/validators.py

```diff
@@ -0,0 +1,62 @@
+from typing import (
+    Type
+)
+
+from evm.chains.base import (
+    BaseChain
+)
+from evm.exceptions import (
+    ValidationError
+)
+from evm.rlp.transactions import (
+    BaseTransaction,
+    BaseTransactionFields
+)
+
+
+class DefaultTransactionValidator():
+    """
+    The :class:`~trinity.tx_pool.validators.DefaultTransactionValidator` class is responsible to
+    decide wether transactions should be relayed to peers or not. This implementation validates
+    transactions against a transaction class inferred from a ``initial_tx_validation_block_number``
+    but will switch to a different one as soon as the tip of the chain uses a more up to date
+    transaction class than the one that corresponds to the ``initial_tx_validation_block_number``.
+    """
+
+    def __init__(self, chain: BaseChain, initial_tx_validation_block_number: int) -> None:
+        self.chain = chain
+        self._initial_tx_class = self._get_tx_class_for_block_number(
+            initial_tx_validation_block_number
+        )
+        self._ordered_tx_classes = [
+            vm_class.get_transaction_class()
+            for _, vm_class in chain.get_vm_configuration()
+        ]
+        self._initial_tx_class_index = self._ordered_tx_classes.index(self._initial_tx_class)
+
+    def __call__(self, transaction: BaseTransactionFields) -> bool:
+
+        transaction_class = self.get_appropriate_tx_class()
+        tx = transaction_class(**transaction.as_dict())
+        try:
+            tx.validate()
+        except ValidationError:
+            return False
+        else:
+            return True
+
+    def get_appropriate_tx_class(self) -> Type[BaseTransaction]:
```

Comment:
This could probably be cached.  Maybe use a time-based caching approach and expire the cache after a few seconds (like 5?).  Probably save us a **lot** of inter-process communication talking back and forth with the chain for every single transaction we want to validate.

Reply:
Sounds like a good idea. Is there a standard approach to this? I've found this one but I'm not sure if it's worth the dependency for this single place of use?

http://cachetools.readthedocs.io/en/latest/#cachetools.func.ttl_cache

Anyway, I will create an issue for this so that I can merge before we get to decide this.